torchmultimodal/modules/layers/transformer_decoder.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import nn
+from torch.utils.checkpoint import checkpoint
+
+from torchmultimodal.modules.layers.attention import FullAttention, MultiHeadAttention
+from torchmultimodal.modules.layers.mlp import MLP
+
+
+class SiLU(nn.Module):
+    r"""Sigmoind Linear Unit
+
+    .. math:: \text{SiLU}(x) = x * \sigma(1.702 * x)
+
+    where :math:`\sigma(x)` is the cumulative distribution function for Logistic Distribution.
+
+    Approximation of the exact GeLU implemented by pytorch for greater forward speed.
+    Reference: `"Gaussian error linear units"<https://arxiv.org/pdf/1606.08415.pdf>`_.
+    """
+
+    def forward(self, x):
+        return torch.sigmoid(1.702 * x) * x
+
+
+class RightShift(nn.Module):
+    def __init__(self, embedding_dim):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.sos = nn.Parameter(
+            torch.FloatTensor(embedding_dim).normal_(std=0.02), requires_grad=True
+        )
+
+    def forward(self, x, decode_step=None):
+        if decode_step is not None and decode_step > 0:
+            return x
+
+        x_shape = list(x.shape)
+        x = x.flatten(start_dim=1, end_dim=-2)  # (batch, seq_len, embedding_dim)
+        sos = (
+            torch.ones(x_shape[0], 1, self.embedding_dim, dtype=torch.float32).to(
+                self.sos
+            )
+            * self.sos
+        )  # (batch, 1, embedding_dim)
+        sos = sos.type_as(x)
+        # Shift one unit to the right along dim ``seq_len``
+        x = torch.cat([sos, x[:, :-1, :]], axis=1)  # (batch, seq_len, embedding_dim)
+        x = x.view(*x_shape)
+        return x
+
+
+class MLPBlock(MLP):
```

Comment:
Is this based on [this layer](https://github.com/mugen-org/Baseline/blob/7961b537ba49fa5bd1ada099e0f6067894254cc0/lib/models/gpt/attention.py#L101)? If so, can you just use our MLP directly?